miri-script/src/commands.rs

```diff
@@ -0,0 +1,711 @@
+struct Extras {
+    sysroot: PathBuf,
+    libdir: PathBuf,
+}
+
+impl Extras {
+    fn new(toolchain: &str) -> Result<Self> {
+        // Determine toolchain properties.
+        let toolchain_prop = &format!("+{toolchain}");
+        let target_output = cmd("rustc", [&toolchain_prop, "--version", "--verbose"]).read()?;
+        let host_line = target_output
+            .lines()
+            .find(|line| line.starts_with("host:"))
+            .expect("rustc output must contain host info");
+        let target = host_line.split(' ').nth(1).expect("host line does not contain host info");
+
+        let sysroot = cmd("rustc", [&toolchain_prop, "--print", "sysroot"]).read()?;
+
+        let sysroot = PathBuf::from(sysroot);
+        let libdir = path!(sysroot / "lib" / "rustlib" / target / "lib");
+        Ok(Self { libdir, sysroot })
+    }
+}
+
+impl<'a> MiriRunner<'a> {
+    pub(super) fn run(command: &'a super::Subcommands) -> Result<()> {
+        let miri_dir = detect_miri_dir()?;
+        let active_toolchain = get_current_toolchain(&miri_dir)?;
+        let config = command.get_auto_config(&miri_dir);
+        let cargo_extra_flags = std::env::var("CARGO_EXTRA_FLAGS").unwrap_or_default();
+        let extra_flags = BTreeMap::new();
+
+        let mut ret = Self {
+            miri_dir,
+            active_toolchain,
+            command,
+            extra_flags,
+            extra_vars: Default::default(),
+            cargo_extra_flags,
+        };
+        if let Some(config) = config {
+            ret.extra_flags.insert("MIRI_AUTO_OPS".into(), "42".into());
+            if config.toolchain {
+                ret.toolchain(&[])?;
+            }
+            if config.fmt {
+                ret.fmt(&[])?;
+            }
+            if config.clippy {
+                ret.clippy(&["--".to_owned(), "-D".to_owned(), "warnings".to_owned()])?;
+            }
+            if config.toolchain {
+                // Let's make sure to actually use that toolchain, too.
+                // We do it after running all the auto things though to use a detected toolchain.
+                ret.active_toolchain = "miri".to_owned();
+            }
+        }
+        let extra_vars = Extras::new(&ret.active_toolchain)?;
+        if !extra_vars.libdir.exists() {
+            println!("Something went wrong determining the library dir.");
+            println!("I got {} but that does not exist.", extra_vars.libdir.display());
+            println!("Please report a bug at https://github.com/rust-lang/miri/issues.");
+            std::process::exit(2);
+        }
+        ret.extra_vars = extra_vars;
+        // Share target dir between `miri` and `cargo-miri`.
+        let target_dir = std::env::var("CARGO_TARGET_DIR")
+            .ok()
+            .filter(|val| !val.is_empty())
+            .unwrap_or_else(|| {
+                let target_dir = path!(ret.miri_dir / "target");
+                target_dir.to_string_lossy().to_string()
```

Comment:
Same comment on all the use of `to_string_lossy` as I had on pass-through CLI arguments. Perhaps that means we also need `env::var_os`?